### ISSUE TITLE
Rule proposal: Use named capture groups for regex

### ISSUE THREAD

**Issue by sindresorhus** [eslint/eslint#11381]
**Please describe what the rule should do:**

Enforce using [named capture groups](http://2ality.com/2017/05/regexp-named-capture-groups.html) for regexes instead numbered capture groups. Named capture groups make regexes more readable and they can also prevent possible off-by-one errors.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[x] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
/([0-9]{4})-([0-9]{2})-([0-9]{2})/;
```

```js
/Chapter (\d+)\.\d*/;
```

For reference, the first example would look like this with named capture groups:

```js
/(?<year>[0-9]{4})-(?<month>[0-9]{2})-(?<day>[0-9]{2})/;
```

**Why should this rule be included in ESLint (instead of a plugin)?**

It's a native JS feature and ESLint already has some other regex-related rules.

**Are you willing to submit a pull request to implement this rule?**

No

**Comment by platinumazure** [comment]
I'm in favor of this in general, although I think I would characterize this as a suggestion rather than a problem (i.e., the rule doesn't report a known/very likely actual bug, but rather can help avoid a type of bug that might or might not occur in a project).

**Comment by sindresorhus** [comment]
@platinumazure Yeah, I was not sure about that. It could potentially catch a problem, but you're right, the main intention is readability. Changed.

**Comment by ilyavolodin** [comment]
Anyone wants to champion this rule? It has enough up-votes to be accepted, just need a champion.

**Comment by mysticatea** [comment]
✋

**Comment by not-an-aardvark** [comment]
Sometimes, people use parentheses to indicate grouping in a regex without using it as a capture group, e.g.

```js
const regex = /(ab|cd|ef)+/;
```

Would this report an error? It could be replaced with `/(?:ab|cd|ef)+/`, although maybe some users would think using a non-capturing group is unnecessary. (Numbered capture groups are generally only a readability issue if the groups are actually accessed on a match.)

(This comment is cross-posted from https://github.com/eslint/eslint/pull/11392#discussion_r258767237 for more visibility.)

**Comment by sindresorhus** [comment]
> Would this report an error?

IMHO, yes, but if you think it's a big problem, there could be an option to opt-out of that.
